### PR TITLE
Add error handling to Get-InstalledPSResource and Find-PSResource

### LINKED ISSUE #699: Find/Get-PSResource does not write error when it cannot find resource
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PowerShellGet)
- [X] Search the existing issues.

### Steps to reproduce

If you use `Find-PSResource` or `Get-PSResource` on a non-existent resource no error will be output.

### Expected behavior

```console
Error is written
```

### Actual behavior

```console
No error
```

### Error details

```console
N/A
```

### Environment data

```powershell
Name          Version
----          -------
PowerShellGet 3.0.14

Name                           Value
----                           -----
PSVersion                      7.2.5
PSEdition                      Core
GitCommitId                    7.2.5
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by alerickson**
# PR Summary
Add better error messaging to `Find-PSResource` when resource is not found.  
Add error handling to `Get-InstalledPSResource`.

## PR Context
Resolves #699 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/GetInstalledPSResource.cs [hunk 1]
```diff
@@ -138,14 +139,56 @@ protected override void ProcessRecord()
...
+                    else
+                    {
+                        pkgsNotFound.Add(name);
+                    }
+                }
+            }
+
+            if (pkgsNotFound.Count > 0)
+            {
+                WriteError(new ErrorRecord(
```
Comment (adityapatwardhan) [inline src/code/GetInstalledPSResource.cs:187]: One caveat here. Typically in PowerShell if a wildcard is specified and the item is not found, you dont get an error. So maybe the following code above is not required:

```csharp
if (!foundWildcardMatch)
{
    pkgsNotFound.Add(name);
}
```

File: test/GetInstalledPSResource/GetInstalledPSResource.Tests.ps1 [hunk 2]
```diff
@@ -127,14 +127,19 @@ $testCases =
...
-        $res = Get-InstalledPSResource -Name $testScriptName -Version "3.0.0"
+        $res = Get-InstalledPSResource -Name $testScriptName -Version "3.0.0" -ErrorAction SilentlyContinue
         $res | Should -BeNullOrEmpty
         $res = Get-InstalledPSResource -Name $testScriptName -Version "3.0.0-alpha"
         $res.Name | Should -Be $testScriptName
         $res.Version | Should -Be "3.0.0"
         $res.Prerelease | Should -Be "alpha"
     }
 
+    It "Throw package not found error when searching for module that does not exist" {
```
Comment (adityapatwardhan) [inline test/GetInstalledPSResource/GetInstalledPSResource.Tests.ps1:138]: Add a test for wildcard pattern not found and no error is thrown

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/code/FindHelper.cs [hunk 3]
```diff
@@ -439,8 +440,10 @@ private IEnumerable<PSResourceInfo> SearchByNames(ServerApiCall currentServer, R
                             if (currentResult.exception != null && !currentResult.exception.Message.Equals(string.Empty))
                             {
                                 // write warning?
```
Comment (anamnavi) [inline src/code/FindHelper.cs:442]:
```suggestion
```

File: src/code/FindHelper.cs [hunk 4]
```diff
@@ -460,14 +463,15 @@ private IEnumerable<PSResourceInfo> SearchByNames(ServerApiCall currentServer, R
...
                         if (_tag.Length == 0)
                         {
                             responses = currentServer.FindName(pkgName, _prerelease, _type, out errRecord);
                         }
                         else
                         {
                             responses = currentServer.FindNameWithTag(pkgName, _tag, _prerelease, _type, out errRecord);
-                            string tagsAsString = String.Join(", ", _tag);
+                            tagsAsString = String.Join(", ", _tag);
                             tagMsg = $" and Tags {tagsAsString}";
```
Comment (anamnavi) [inline src/code/FindHelper.cs:475]:
```suggestion
```

File: src/code/FindHelper.cs [hunk 5]
```diff
@@ -417,14 +417,15 @@ private IEnumerable<PSResourceInfo> SearchByNames(ServerApiCall currentServer, R
...
                         if (_tag.Length == 0)
                         {
                             responses = currentServer.FindNameGlobbing(pkgName, _prerelease, _type, out errRecord);
                         }
                         else
                         {
                             responses = currentServer.FindNameGlobbingWithTag(pkgName, _tag, _prerelease, _type, out errRecord);
-                            string tagsAsString = String.Join(", ", _tag);
+                            tagsAsString = String.Join(", ", _tag);
                             tagMsg = $" and Tags {tagsAsString}";
```
Comment (anamnavi) [inline src/code/FindHelper.cs:429]:
```suggestion
```

**Review by anamnavi — CHANGES_REQUESTED** [review]
LGTM, just some minor changes in Find-PSResource.
You don't need the `tagsMsg` variable anymore. Initially I had intended it to contain the tag part of the message if we did a tag type of search. And if we didn't then the string would be empty. And then the error message should have been like `package with name X and {tagMsg} was not found` but the error message contained the words `...and tags {tagMsg}` and it defeated the point of this. So long story short since we're doing it differently now we can just get rid of that `tagMsg` variable :)

**Review by adityapatwardhan — APPROVED** [review]

**Review by anamnavi — APPROVED** [review]

**Pull request merged**